server/meca/article.test.js

```diff
@@ -0,0 +1,38 @@
+const generateXml = require('./article')
+const replaySetup = require('../../test/helpers/replay-setup')
+const sampleManuscript = require('./article.test.data')
+
+replaySetup('success')
+
+const snippets = [
+  ['doctype', '<!DOCTYPE article SYSTEM "JATS-journalpublishing1.dtd">'],
+  [
+    'article id',
+    '<article-id pub-id-type="manuscript">604e06ca-882d-4b5b-a147-e016893e60e9</article-id>',
+  ],
+  ['article type', '<subject>research-article</subject>'],
+  ['article type id', '<subject>5</subject>'],
+  ['subject area ID', '<subject>chromosomes-gene-expression</subject>'],
+  ['title', '<article-title>Test Manuscript</article-title>'],
+  ['previously discussed', '<meta-value>Talked to bob about it</meta-value>'],
+  [
+    'created by',
+    '<meta-value>6d8cd1ce-15b6-46c1-b901-bc91598c8f2d</meta-value>',
+  ],
+  ['author email', '<email>elife@mailinator.com</email>'],
+  [
+    'related article',
+    `<related-article related-article-type="companion">
+        <article-title>Another</article-title>
+      </related-article>`,
+  ],
+]
+
+describe('Article XML generator', () => {
+  describe('sample output', () => {
+    test.each(snippets)('includes %s', async (name, snippet) => {
+      const xml = await generateXml(sampleManuscript)
+      expect(xml).toMatch(snippet)
```

Comment:
These tests are a great smoke test to ensure that the XML is correct where it matters, but I'm worried that given the complexity of the logic that generates the XML, we may have some cases we haven't covered.

I don't think it's feasible to address this given the state of the code above unless we use a code coverage tool against this,  but instead it's worth refactoring the logic to make it easier to identify edge cases as well as making it easier to maintain.

Reply:
Yes, certainly there are some parts of the output which are not asserted against. I don't think a test coverage tool would help here though (jest reports 100% coverage).

An alternative is to assert on the full XML output in one go. The default test failure output would be quite hard to read but there's a [custom matcher](https://www.npmjs.com/package/jest-xml-matcher) which could help.